compiler/compiler_test.go

```diff
@@ -405,6 +406,33 @@ func TestLengthParenthesizingIssue841(t *testing.T) {
 	}
 }
 
+func TestArchiveSelectionAfterSerialization(t *testing.T) {
+	src := `
+		package main
+		type Foo interface{ int | string }
+
+		type Bar[T Foo] struct{ v T }
+		func (b Bar[T]) Baz() { println(b.v) }
+
+		var ghost = Bar[int]{v: 7} // unused
+
+		func main() {
+			println("do nothing")
+		}`
+	srcFiles := []srctesting.Source{{Name: `main.go`, Contents: []byte(src)}}
+	root := srctesting.ParseSources(t, srcFiles, nil)
+	rootPath := root.PkgPath
+	origArchives := compileProject(t, root, false)
+	readArchives := reloadCompiledProject(t, origArchives, rootPath)
+
+	origJS := renderPackage(t, origArchives[rootPath], false)
+	readJS := renderPackage(t, readArchives[rootPath], false)
```

Comment:
Can we just diff the archive slices? I'm reasonably certain cmp.Diff() can handle unexported fields if you pass it the right options.

Reply:
Sure, I hadn't tried that and just used JS because I was basing this test off of the `TestOrder` test.

I also wanted to make sure decl DCE selection was working correctly since I was also fixing that. As long as `cmp.Diff` also checks unexpected data then it should be fine.

Reply:
My thinking was that if the test is aimed at testing the invariant of `ReadArchive(WriteArchive(a)) == a`, it's best to test it directly. Adding code generation into the mix can cause additional complexity.

The tests for DCE you added in the other PR seemed comprehensive enough to me.

Reply:
Oops, didn't mean "unexpected", I meant "unexported". So I gave this a try and apparently is does complain with unexported data. I added the following into the test prior to the `renderPackage`s:

```Go
if diff := cmp.Diff(origArchives, readArchives); diff != `` {
	t.Errorf("the reloaded archives didn't match:\n%s", diff)
}
```

and `cmp.Diff` panicked with

```Plain
panic: cannot handle unexported field at {map[string]*compiler.Archive}["command-line-arguments"].Package.path:
		"go/types".Package
	consider using a custom Comparer; if you control the implementation of type, you can also consider using an Exporter, AllowUnexported, or cmpopts.IgnoreUnexported
```

The `AllowUnexported` option requires all the unexported types to be passed in, but trying to pass in everything in "go/types" eventually reaches `types.object` that I can't pass into the option. The [`Exporter`](https://pkg.go.dev/github.com/google/go-cmp/cmp#Exporter) option does make it possible to test all unexported types (`cmp.Exporter(func(_ reflect.Type) bool { return true })`) but the documentation recommend against that and we're getting diffs in things like `pos`, `color_`, `id`, etc.

I think at least for right now it is easiest to just have the JS rendered as a way to confirm the important information is being reloaded and anything else not loaded identically will not affect the results.

Reply:
That's fair, let's keep it as is then.